src/hb-ot-layout-gsubgpos.hh

```diff
@@ -3678,57 +3678,60 @@ struct GSUBGPOS
                                 const hb_set_t *feature_indices,
                                 hb_map_t *duplicate_feature_map /* OUT */) const
   {
+    hb_hashmap_t<hb_tag_t, hb_set_t *, (unsigned)-1, nullptr> tag_features;
+    
     //find out duplicate features after subset
-    unsigned prev = 0xFFFFu;
     for (unsigned i : feature_indices->iter ())
     {
-      if (prev == 0xFFFFu)
-      {
-        duplicate_feature_map->set (i, i);
-        prev = i;
-        continue;
-      }
-
       hb_tag_t t = get_feature_tag (i);
-      hb_tag_t prev_t = get_feature_tag (prev);
-      if (t != prev_t)
+      if (!tag_features.has (t))
       {
+        hb_set_t *same_tag_features = hb_set_create ();
+        same_tag_features->add (i);
+        tag_features.set (t, same_tag_features);
         duplicate_feature_map->set (i, i);
-        prev = i;
         continue;
       }
 
-      const Feature& f = get_feature (i);
-      const Feature& prev_f = get_feature (prev);
-
-      auto f_iter =
-      + hb_iter (f.lookupIndex)
-      | hb_filter (lookup_indices)
-      ;
-
-      auto prev_iter =
-      + hb_iter (prev_f.lookupIndex)
-      | hb_filter (lookup_indices)
-      ;
+      hb_set_t *same_tag_features = tag_features.get (t);
+      bool found = false;
 
-      if (f_iter.len () != prev_iter.len ())
+      for (unsigned other_f_index : same_tag_features->iter ())
       {
-        duplicate_feature_map->set (i, i);
-        prev = i;
-        continue;
+        const Feature& f = get_feature (i);
+        const Feature& other_f = get_feature (other_f_index);
+
+        auto f_iter =
+        + hb_iter (f.lookupIndex)
+        | hb_filter (lookup_indices)
+        ;
+  
+        auto other_f_iter =
+        + hb_iter (other_f.lookupIndex)
+        | hb_filter (lookup_indices)
+        ;
+  
+        if (f_iter.len () != other_f_iter.len ())
+          continue;
+
+        bool is_equal = true;
+        for (auto _ : + hb_zip (f_iter, other_f_iter))
+          if (_.first != _.second) { is_equal = false; break; }
```

Comment:
For a slightly more compact way of checking iterator equality see: https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-cmap-table.hh#L1488 it skips the len() comparison (which ends up being more efficient since calling .len() triggers a full iteration)